include/deal.II/matrix_free/tools.h

```diff
@@ -27,8 +30,369 @@ DEAL_II_NAMESPACE_OPEN
  * evaluation.
  */
 namespace MatrixFreeTools
-{} // namespace MatrixFreeTools
+{
+  namespace internal
+  {
+    template <typename Number>
+    struct LocalCSR
+    {
+      std::vector<unsigned int> row_lid_to_gid;
+      std::vector<unsigned int> row{0};
+      std::vector<unsigned int> col;
+      std::vector<Number>       val;
+    };
+  } // namespace internal
 
+  template <int dim,
+            int fe_degree,
+            int n_q_points_1d,
+            int n_components,
+            typename Number,
+            typename VectorizedArrayType>
+  void
+  compute_diagonal(
+    const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+    FEEvaluation<dim,
+                 fe_degree,
+                 n_q_points_1d,
+                 n_components,
+                 Number,
+                 VectorizedArrayType> &                 phi,
+    LinearAlgebra::distributed::Vector<Number> &        diagonal_global,
+    const std::function<void()> &                       local_vmult)
```

Comment:
Can we add input argument of an unsigned int here and pass the cell index below? I need to query a coefficient table in my `local_vmult()`. 
```suggestion
    const std::function<void(const unsigned int)>&                       local_vmult)
```

Reply:
@tcclevenger Sorry for the late reply! Thanks for testing the code in ASPECT. I don't think we need the additional argument. A new function has been introduced recently, s.t. it is possible to get the index from `FEEvaluation`:

https://github.com/dealii/dealii/blob/040c6ad7d46f1291ab5c038b70cf1da4fb22b006/include/deal.II/matrix_free/fe_evaluation.h#L237-L241

That should also work, shouldn't it?